### PR TITLE
[GPU] onednn is not chosen for 3d conv

### PR DESCRIPTION
**Pull request by sungeunk**
Signed-off-by: Kim,SungEun <sungeun.kim@intel.com>

### Tickets:
 - 74548

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp [hunk 1]
```diff
@@ -103,6 +103,13 @@ void remove_redundant_reorders::run(program& p) {
             if (!all_users_fuse)
                 continue;
 
+            // keep reorder(bs_fs_zyx_bsv8_fsv2) before first conv(shallow feature)
+            if (output_layout.format == format::bs_fs_zyx_bsv8_fsv2) {
```
Comment (sungeunk) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:107]: move to can_fuse_reorder

Comment (sungeunk) [inline src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:107]: move to can_fuse_reorder

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 2]
```diff
@@ -1354,6 +1390,16 @@ impl_types layout_optimizer::get_preferred_impl_type(program_node& node, format
             impl_candidate = impl_types::ocl;
         }
 
+        // fallback cldnn if onednn pooling's layout is b_fs_zyx_fsv32 and i8.
+        // It filed this issue to onednn team. (MFDNN-7793)
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1394]: Please do not mention ticket id here.

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 3]
```diff
@@ -1551,7 +1597,10 @@ format layout_optimizer::get_preferred_format(program_node& node) {
         } else if (use_onednn_impls && data_type_traits::is_i8_u8(layout.data_type) &&
             needs_all_usr_onednn_small_ic_to_blocked(node)) {
             // All user nodes are convolutions which satisfy options for onednn first conv
-            expected = format::bs_fs_yx_bsv8_fsv4;
+            if (layout.format.spatial_num() == 2)
+                expected = format::bs_fs_yx_bsv8_fsv4;
+            else if (layout.format.spatial_num() == 3)
+                expected = format::bs_fs_zyx_bsv8_fsv4;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:1603]: What about defining a helper function such as below? I think this helper function will make the code much readable..
`expected = get_layout(format::bs_fs_yx_bsv8_fsv4, layout.format.spatial_num())`

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/pooling_gpu_int8_ref.cl [hunk 4]
```diff
@@ -30,7 +30,7 @@ KERNEL(pooling_gpu_int8_ref)(
 #endif
 )
 {
-#if OUTPUT_LAYOUT_BFYX  || OUTPUT_LAYOUT_BYXF || OUTPUT_LAYOUT_B_FS_YX_FSV4 || OUTPUT_LAYOUT_BFZYX
+#if OUTPUT_LAYOUT_BFYX  || OUTPUT_LAYOUT_BYXF || OUTPUT_LAYOUT_B_FS_YX_FSV4 || OUTPUT_LAYOUT_BFZYX || OUTPUT_LAYOUT_BS_FS_ZYX_BSV32_FSV32
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/pooling_gpu_int8_ref.cl:33]: `bs_fs_zyx_bsv16_fsv16` is missing here.

File: src/plugins/intel_gpu/src/kernel_selector/core/common/kernel_selector_utils.cpp [hunk 5]
```diff
@@ -265,7 +265,8 @@ std::vector<size_t> GetOptimalLocalWorkGroupSizes(std::vector<size_t> gws, const
                               output_layout == DataLayout::b_fs_yx_fsv16 || output_layout == DataLayout::b_fs_zyx_fsv16 ||
                               output_layout == DataLayout::b_fs_yx_fsv32 || output_layout == DataLayout::b_fs_zyx_fsv32;
 
-    auto blocked_bsv_fsv_layout = output_layout == DataLayout::bs_fs_yx_bsv16_fsv16 || output_layout == DataLayout::bs_fs_zyx_bsv16_fsv16;
+    auto blocked_bsv_fsv_layout = output_layout == DataLayout::bs_fs_yx_bsv16_fsv16 || output_layout == DataLayout::bs_fs_zyx_bsv16_fsv16 ||
+                                  output_layout == DataLayout::bs_fs_zyx_bsv16_fsv32 || output_layout == DataLayout::bs_fs_zyx_bsv32_fsv32;
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/common/kernel_selector_utils.cpp:269]: why bsv32_fsv16 is missing here?

File: src/plugins/intel_gpu/src/kernel_selector/core/common/jitter.cpp [hunk 6]
```diff
@@ -398,6 +398,18 @@ JitDefinitions DataTensorJitConstant::GetDefinitions() const {
                 index_func_val = "GET_DATA_B_FS_ZYX_FSV16_INDEX(" + _name + ", b, f, z, y, x)";
                 raw_index_func_val = "GET_DATA_B_FS_ZYX_FSV16_INDEX(" + _name + ", b, f, z, y, x)";
                 safe_index_func_val = "GET_DATA_B_FS_ZYX_FSV16_INDEX_SAFE(" + _name + ", b, f, z, y, x)";
+            } else if (layout == DataLayout::bs_fs_zyx_bsv32_fsv32) {
+                index_func_val = "GET_DATA_BS_FS_ZYX_BSV32_FSV32_INDEX(" + _name + ", b, f, z, y, x)";
```
Comment (isanghao) [inline src/plugins/intel_gpu/src/kernel_selector/core/common/jitter.cpp:402]: Can we use `toString` function here?

File: src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp [hunk 7]
```diff
@@ -489,7 +489,9 @@ struct tensor {
 
             // skip z for the formats that do not have it
             if (((new_fmt != format::bfzyx && new_fmt != format::b_fs_zyx_fsv16 && new_fmt != format::b_fs_zyx_fsv32 &&
-                  new_fmt != format::bfwzyx && new_fmt != format::bs_fs_zyx_bsv16_fsv16)) && (c == 'z')) {
+                  new_fmt != format::bfwzyx && new_fmt != format::bs_fs_zyx_bsv16_fsv16 && new_fmt != format::bs_fs_zyx_bsv16_fsv32 &&
+                  new_fmt != format::bs_fs_zyx_bsv32_fsv16 && new_fmt != format::bs_fs_zyx_bsv32_fsv32 &&
+                  new_fmt != format::bs_fs_zyx_bsv8_fsv2 && new_fmt != format::bs_fs_zyx_bsv8_fsv4)) && (c == 'z')) {
```
[the remark this answers is not in the document]
Comment (sungeunk) [reply src/plugins/intel_gpu/include/intel_gpu/runtime/tensor.hpp:494]: I guess 'skip z' means the z value will be multiply to next tensor.

File: src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp [hunk 8]
```diff
@@ -385,6 +392,28 @@ void remove_redundant_reorders::run(program& p) {
         p.remove_if_dangling(node);
     }
 
+    // This pass removed reorder if it is between quantize and convolution.
```
[the remark this answers is not in the document]
Comment (sungeunk) [reply src/plugins/intel_gpu/src/graph/graph_optimizer/remove_redundant_reorders.cpp:388]: Fusing logics are different.
[onenote:different_fusing_after_can_fuse_reorder](https://intel.sharepoint.com/sites/AVAopt/_layouts/OneNote.aspx?id=%2Fsites%2FAVAopt%2FSiteAssets%2FAVA%20opt%20Notebook&wd=target%28GPU-Dev1.one%7CFFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F%2Fcode%20review%3A%20remove_redundant_reorders%7CCC2BF377-2896-41DF-A181-88327D01153B%2F%29
onenote:https://intel.sharepoint.com/sites/AVAopt/SiteAssets/AVA%20opt%20Notebook/GPU-Dev1.one#code%20review%20remove_redundant_reorders&section-id={FFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F}&page-id={CC2BF377-2896-41DF-A181-88327D01153B}&end)

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/pooling_gpu_int8_ref.cl (hunk 4 above)
Comment (sungeunk) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/pooling_gpu_int8_ref.cl]: add it

File: src/plugins/intel_gpu/src/kernel_selector/core/common/kernel_selector_utils.cpp (hunk 5 above)
Comment (sungeunk) [reply src/plugins/intel_gpu/src/kernel_selector/core/common/kernel_selector_utils.cpp]: add it

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp (hunk 2 above)
Comment (sungeunk) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp]: remove it

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 9]
```diff
@@ -421,7 +412,7 @@ bool layout_optimizer::can_fuse_reorder_to_prev(program_node& prev, program_node
 
     if (prev.is_type<quantize>() &&
         (fmt_next == format::b_fs_yx_fsv4 || fmt_next == format::b_fs_zyx_fsv32 || (fmt_next == format::b_fs_yx_fsv32 && !use_onednn_impls) ||
-         fmt_next == format::b_fs_yx_fsv16 || fmt_next == format::b_fs_zyx_fsv16 || fmt_next == format::bs_fs_yx_bsv16_fsv16))
+         fmt_next == format::b_fs_yx_fsv16 || fmt_next == format::bs_fs_yx_bsv16_fsv16))
```
[the remark this answers is not in the document]
Comment (sungeunk) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:421]: Fusing logics are different after calling can_fuse_reorder_to_prev.
[onenote:different_fusing_after_can_fuse_reorder](https://intel.sharepoint.com/sites/AVAopt/_layouts/OneNote.aspx?id=%2Fsites%2FAVAopt%2FSiteAssets%2FAVA%20opt%20Notebook&wd=target%28GPU-Dev1.one%7CFFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F%2Fcode%20review%3A%20remove_redundant_reorders%7CCC2BF377-2896-41DF-A181-88327D01153B%2F%29onenote:https://intel.sharepoint.com/sites/AVAopt/SiteAssets/AVA%20opt%20Notebook/GPU-Dev1.one#code%20review%20remove_redundant_reorders&section-id=%7BFFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F%7D&page-id=%7BCC2BF377-2896-41DF-A181-88327D01153B%7D&end)

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 10]
```diff
@@ -793,12 +790,15 @@ bool layout_optimizer::needs_onednn_small_ic_to_blocked(format fmt_next, layout&
     auto next_output_layout = node.get_output_layout();
     if (!(prev_output_layout.data_type == next_output_layout.data_type ||
         (prev_output_layout.data_type == data_types::i8 && next_output_layout.data_type == data_types::u8) ||
-        (prev_output_layout.data_type == data_types::u8 && next_output_layout.data_type == data_types::i8)))
+        (prev_output_layout.data_type == data_types::u8 && next_output_layout.data_type == data_types::i8) ||
+        (prev_output_layout.data_type == data_types::i8 && next_output_layout.data_type == data_types::f16)))
```
[the remark this answers is not in the document]
Comment (sungeunk) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:800]: Checking precision could be removed.
[onenote:small_ic_to_blocked](https://intel.sharepoint.com/sites/AVAopt/_layouts/OneNote.aspx?id=%2Fsites%2FAVAopt%2FSiteAssets%2FAVA%20opt%20Notebook&wd=target%28GPU-Dev1.one%7CFFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F%2Fcode%20review%3A%20small_ic_to_blocked%7CBDDDEA05-2B89-4519-9651-89ED52F4369C%2F%29onenote:https://intel.sharepoint.com/sites/AVAopt/SiteAssets/AVA%20opt%20Notebook/GPU-Dev1.one#code%20review%20small_ic_to_blocked&section-id={FFDEBE81-16EC-47EF-8A91-F9F0FD8BD24F}&page-id={BDDDEA05-2B89-4519-9651-89ED52F4369C}&end)

File: src/plugins/intel_gpu/src/graph/layout_optimizer.cpp [hunk 11]
```diff
@@ -793,12 +790,15 @@ bool layout_optimizer::needs_onednn_small_ic_to_blocked(format fmt_next, layout&
...
         return false;
 
     // Target output_layout format
     if (!(fmt_next == format::b_fs_yx_fsv16 || fmt_next == format::bs_fs_yx_bsv32_fsv16 ||
-        fmt_next == format::b_fs_yx_fsv32 || fmt_next == format::bs_fs_yx_bsv32_fsv32))
+        fmt_next == format::b_fs_yx_fsv32 || fmt_next == format::bs_fs_yx_bsv32_fsv32 ||
+        fmt_next == format::bs_fs_zyx_bsv32_fsv16 || fmt_next == format::b_fs_zyx_fsv32 ||
+        fmt_next == format::b_fs_zyx_fsv16 || fmt_next == format::bs_fs_zyx_bsv8_fsv4))
```
[the remark this answers is not in the document]
Comment (sungeunk) [reply src/plugins/intel_gpu/src/graph/layout_optimizer.cpp:807]: bs_fs_zyx_bsv8_fsv4 is unnecessary format here. I'll remove it.

**Review by isanghao — APPROVED** [review]

**Pull request merged**